// @redis/time-series/dist/lib/commands/MGET.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseFilterArgument = exports.parseLatestArgument = void 0;
const helpers_1 = require("./helpers");
function parseLatestArgument(parser, latest) {
    if (latest) {
        parser.push('LATEST');
    }
}
exports.parseLatestArgument = parseLatestArgument;
function parseFilterArgument(parser, filter) {
    parser.push('FILTER');
    parser.pushVariadic(filter);
}
exports.parseFilterArgument = parseFilterArgument;
exports.default = {
    NOT_KEYED_COMMAND: true,
    IS_READ_ONLY: true,
    parseCommand(parser, filter, options) {
        parser.push('TS.MGET');
        parseLatestArgument(parser, options?.LATEST);
        parseFilterArgument(parser, filter);
    },
    transformReply: {
        2(reply, _, typeMapping) {
            return (0, helpers_1.resp2MapToValue)(reply, ([, , sample]) => {
                return {
                    sample: helpers_1.transformSampleReply[2](sample)
                };
            }, typeMapping);
        },
        3(reply) {
            return (0, helpers_1.resp3MapToValue)(reply, ([, sample]) => {
                return {
                    sample: helpers_1.transformSampleReply[3](sample)
                };
            });
        }
    }
};
